src/Microsoft.ApiDesignGuidelines.Analyzers/Core/IdentifiersShouldHaveCorrectSuffix.cs

```diff
@@ -25,24 +29,141 @@ public abstract class IdentifiersShouldHaveCorrectSuffixAnalyzer : DiagnosticAna
+            var interfaceTypeSuffixMap = new Dictionary<INamedTypeSymbol, SuffixInfo>();
+
+            foreach (var tuple in s_baseTypesAndTheirSuffix)
+            {
+                var wellKnownNamedType = context.Compilation.GetTypeByMetadataName(tuple.Item1);
+
+                if (wellKnownNamedType != null && wellKnownNamedType.OriginalDefinition != null)
+                {
+                    // If the type is interface
+                    if (wellKnownNamedType.OriginalDefinition.TypeKind == TypeKind.Interface)
+                    {
+                        interfaceTypeSuffixMap.Add(wellKnownNamedType.OriginalDefinition, SuffixInfo.Create(tuple.Item2, tuple.Item3));
+                    }
+                    else
+                    {
+                        baseTypeSuffixMap.Add(wellKnownNamedType.OriginalDefinition, SuffixInfo.Create(tuple.Item2, tuple.Item3));
+                    }
+                }
+            }
+
+            if (baseTypeSuffixMap.Count > 0 || interfaceTypeSuffixMap.Count > 0)
+            {
+                context.RegisterSymbolAction((saContext) =>
+                {
+                    var namedTypeSymbol = saContext.Symbol as INamedTypeSymbol;
+                    if (namedTypeSymbol.GetResultantVisibility() != SymbolVisibility.Public)
+                    {
+                        return;
+                    }
+
+                    var baseType = namedTypeSymbol.GetBaseTypes().FirstOrDefault(bt => baseTypeSuffixMap.Keys.Contains(bt.OriginalDefinition));
+                    if (baseType != null)
+                    {
+                        var suffixInfo = baseTypeSuffixMap[baseType.OriginalDefinition];
+                        var rule = suffixInfo.CanSuffixBeCollection ? SpecialCollectionRule : DefaultRule;
+                        if ((suffixInfo.CanSuffixBeCollection && !namedTypeSymbol.Name.EndsWith(CollectionString) && !namedTypeSymbol.Name.EndsWith(suffixInfo.Suffix)) ||
+                            (!suffixInfo.CanSuffixBeCollection && !namedTypeSymbol.Name.EndsWith(suffixInfo.Suffix)))
+                        {
+                            saContext.ReportDiagnostic(namedTypeSymbol.CreateDiagnostic(rule, namedTypeSymbol.ToDisplayString(), suffixInfo.Suffix));
+                        }
+
+                        return;
+                    }
+
+                    var implementedInterface = namedTypeSymbol.AllInterfaces.FirstOrDefault(i => interfaceTypeSuffixMap.Keys.Contains(i.OriginalDefinition));
+                    if (implementedInterface != null)
+                    {
+                        var suffixInfo = interfaceTypeSuffixMap[implementedInterface.OriginalDefinition];
+                        if (!namedTypeSymbol.Name.EndsWith(suffixInfo.Suffix))
+                        {
+                            saContext.ReportDiagnostic(namedTypeSymbol.CreateDiagnostic(DefaultRule, namedTypeSymbol.ToDisplayString(), suffixInfo.Suffix));
```

Comment:
What happens when a type implements both a base type and interface (or even multiple interfaces) from the suffix maps? You should probably change the implementation so that first we find the set of base types and interfaces that are from the suffix maps. If non-empty, then report a diagnostic if the symbol name does not end with a suffix that matches the suffix required for any of the types in this set.